runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -3396,6 +3425,664 @@ J9::Z::TreeEvaluator::ArrayCopyBNDCHKEvaluator(TR::Node * node, TR::CodeGenerato
+      }
+   else 
+      return node->getFirstChild();
+   }
+
+/*
+   * Fill in the J9JITWatchedStaticFieldData.fieldAddress, J9JITWatchedStaticFieldData.fieldClass for static field
+   * and J9JITWatchedInstanceFieldData.offset for instance field
+   *
+   */
+
+void generateFillInDataBlockSequenceForUnresolvedField (TR::Node *node, J9Method* owningMethod, TR::S390WritableDataSnippet *dataSnippet, bool isWrite, TR::CodeGenerator *cg)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)(cg->fe());
+   TR::SymbolReference *symRef = node->getSymbolReference();
+   bool is64Bit = TR::Compiler->target.is64Bit();
+   bool isStatic = symRef->getSymbol()->getKind() == TR::Symbol::IsStatic;
+
+   TR_RuntimeHelper helperIndex = isWrite? (isStatic ?  TR_jitResolveStaticFieldSetterDirect: TR_jitResolveFieldSetterDirect):
+                                          (isStatic ?  TR_jitResolveStaticFieldDirect: TR_jitResolveFieldDirect);
+
+   TR::S390HelperLinkage *helperLink = static_cast<TR::S390HelperLinkage*>(cg->getLinkage(runtimeHelperLinkage(helperIndex)));
+
+   intptr_t offsetInDataBlock = isStatic ? offsetof(J9JITWatchedStaticFieldData, fieldAddress): offsetof(J9JITWatchedInstanceFieldData, offset);
+
+   TR::LabelSymbol* startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol* endLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *unresolvedLabel = generateLabelSymbol(cg);
+   startLabel->setStartInternalControlFlow();
+   endLabel->setEndInternalControlFlow();
+
+   // need 2 registers for arguments, 2 for entry point and return address register, and 1 for datablock
+   uint8_t numOfConditions = 5;
+
+   if (isStatic) // needs fieldClassReg
+      numOfConditions++;
+
+   TR::RegisterDependencyConditions  *deps =  generateRegisterDependencyConditions(0, numOfConditions, cg);
+   TR::Register *resultReg = NULL;
+   TR::Register *tempReg = cg->allocateRegister();
+   TR::Register *dataBlockReg = cg->allocateRegister();
+
+   deps->addPostCondition(dataBlockReg, TR::RealRegister::AssignAny);
+
+   generateRILInstruction(cg, TR::InstOpCode::LARL, node, dataBlockReg, dataSnippet);
+   generateRILInstruction(cg, TR::InstOpCode::LGFI, node, tempReg, static_cast<int32_t>(-1));
+
+   generateRXInstruction(cg, is64Bit ? TR::InstOpCode::CG : TR::InstOpCode::C, node, tempReg, generateS390MemoryReference(dataBlockReg, offsetInDataBlock, cg));
+
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BNE, node, endLabel); // branch to end if not set
+   generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, node, startLabel);
+     
+   if (isStatic)
+      {
+      // Fills in J9JITWatchedStaticFieldData.fieldClass
+      TR::Register *fieldClassReg = cg->evaluate(isWrite ? getDirectWrtbarJ9ClassNode(node): node->getFirstChild());
+      generateRXInstruction(cg, TR::InstOpCode::getStoreOpCode(), node, fieldClassReg, generateS390MemoryReference(dataBlockReg, (int32_t)(offsetof(J9JITWatchedStaticFieldData, fieldClass)), cg));
```

Comment:
I don't think the `(int32_t)` cast is necessary.

Reply:
https://github.com/eclipse/openj9/pull/4743/commits/7c11524271e411af46d035ba51110dc0b0bcee58